server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/AtMostFailureRetries.java

```diff
@@ -0,0 +1,59 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.matchers;
+
+import java.util.Collection;
+import java.util.Optional;
+
+import javax.mail.MessagingException;
+
+import org.apache.james.core.MailAddress;
+import org.apache.mailet.Attribute;
+import org.apache.mailet.AttributeName;
+import org.apache.mailet.AttributeUtils;
+import org.apache.mailet.AttributeValue;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMatcher;
+import org.apache.mailet.base.MailetUtil;
+
+import com.google.common.collect.ImmutableList;
+
+public class AtMostFailureRetries extends GenericMatcher {
+    static final AttributeName MAX_FAILURE_RETRIES = AttributeName.of("MAX_FAILURE_RETRIES");
+    private Integer maxFailureRetries;
+
+    @Override
+    public void init() throws MessagingException {
+        Integer maxFailureRetries = MailetUtil.getInitParameterAsStrictlyPositiveInteger(getCondition());
+        this.maxFailureRetries = maxFailureRetries;
+    }
+
+    @Override
+    public Collection<MailAddress> match(Mail mail) throws MessagingException {
+        return AttributeUtils.getValueAndCastFromMail(mail, MAX_FAILURE_RETRIES, Integer.class)
+            .or(() -> Optional.of(-1))
```

Comment:
Hm because in this case attribute doesn't exist yet. When we set the attribute after it adds +1 (thus setting it to 0 for the mail). Because I think the first time you send a mail, you did not retry yet? I might be wrong or confused though